File: AST.py
from enum import Enum
from collections import namedtuple

class AST:
    def __repr__(self):
        return "{name}({data})".format(name=self.__class__.__name__,
            data=", ".join("{}={}".format(name, getattr(self, name))
                for name in self.__slots__))

operator = Enum('operator', 'Add Sub Mult Div Eq NotEq Lt LtE Gt GtE')

unaryop = Enum('unaryop', 'UAdd USub')

expr_context = Enum('expr_context', 'Load Store')

basic_type = Enum('basic_type', 'Int Character Void')

class Program(AST):
    __slots__ = ('decls')

    def __init__(self, decls):
        self.decls = decls


class ConstDecl(AST):
    __slots__ = ('type', 'name', 'value', 'lineno', 'col_offset')

    def __init__(self, type, name, value, lineno, col_offset):
        self.type = type
        self.name = name
        self.value = value
        self.lineno = lineno
        self.col_offset = col_offset


class VarDecl(AST):
    __slots__ = ('type', 'name', 'lineno', 'col_offset')

    def __init__(self, type, name, lineno, col_offset):
        self.type = type
        self.name = name
        self.lineno = lineno
        self.col_offset = col_offset


class FunctionDef(AST):
    __slots__ = ('return_type', 'name', 'args', 'decls', 'body', 'lineno',
                 'col_offset')

    def __init__(self, return_type, name, args, decls, body, lineno,
                 col_offset):
        self.return_type = return_type
        self.name = name
        self.args = args
        self.decls = decls
        self.body = body
        self.lineno = lineno
        self.col_offset = col_offset


arg = namedtuple('arg', 'type name lineno col_offset')

class Read(AST):
    __slots__ = ('names', 'lineno', 'col_offset')

    def __init__(self, names, lineno, col_offset):
        self.names = names
        self.lineno = lineno
        self.col_offset = col_offset


class Write(AST):
    __slots__ = ('str', 'value', 'lineno', 'col_offset')

    def __init__(self, str, value, lineno, col_offset):
        self.str = str
        self.value = value
        self.lineno = lineno
        self.col_offset = col_offset


class Assign(AST):
    __slots__ = ('target', 'value', 'lineno', 'col_offset')

    def __init__(self, target, value, lineno, col_offset):
        self.target = target
        self.value = value
        self.lineno = lineno
        self.col_offset = col_offset


class For(AST):
    __slots__ = ('initial', 'condition', 'step', 'body', 'lineno', 'col_offset')

    def __init__(self, initial, condition, step, body, lineno, col_offset):
        self.initial = initial
        self.condition = condition
        self.step = step
        self.body = body
        self.lineno = lineno
        self.col_offset = col_offset


class While(AST):
    __slots__ = ('condition', 'body', 'lineno', 'col_offset')

    def __init__(self, condition, body, lineno, col_offset):
        self.condition = condition
        self.body = body
        self.lineno = lineno
        self.col_offset = col_offset


class Switch(AST):
    __slots__ = ('value', 'labels', 'default_', 'lineno', 'col_offset')

    def __init__(self, value, labels, default_, lineno, col_offset):
        self.value = value
        self.labels = labels
        self.default_ = default_
        self.lineno = lineno
        self.col_offset = col_offset


class Return(AST):
    __slots__ = ('value', 'lineno', 'col_offset')

    def __init__(self, value, lineno, col_offset):
        self.value = value
        self.lineno = lineno
        self.col_offset = col_offset


class If(AST):
    __slots__ = ('test', 'body', 'orelse', 'lineno', 'col_offset')

    def __init__(self, test, body, orelse, lineno, col_offset):
        self.test = test
        self.body = body
        self.orelse = orelse
        self.lineno = lineno
        self.col_offset = col_offset


class Expr(AST):
    __slots__ = ('value', 'lineno', 'col_offset')

    def __init__(self, value, lineno, col_offset):
        self.value = value
        self.lineno = lineno
        self.col_offset = col_offset


label_stmt = namedtuple('label_stmt', 'value stmt lineno col_offset')

class BinOp(AST):
    __slots__ = ('left', 'op', 'right', 'lineno', 'col_offset')

    def __init__(self, left, op, right, lineno, col_offset):
        self.left = left
        self.op = op
        self.right = right
        self.lineno = lineno
        self.col_offset = col_offset


class UnaryOp(AST):
    __slots__ = ('op', 'operand', 'lineno', 'col_offset')

    def __init__(self, op, operand, lineno, col_offset):
        self.op = op
        self.operand = operand
        self.lineno = lineno
        self.col_offset = col_offset


class Call(AST):
    __slots__ = ('func', 'args', 'lineno', 'col_offset')

    def __init__(self, func, args, lineno, col_offset):
        self.func = func
        self.args = args
        self.lineno = lineno
        self.col_offset = col_offset


class Num(AST):
    __slots__ = ('n', 'lineno', 'col_offset')

    def __init__(self, n, lineno, col_offset):
        self.n = n
        self.lineno = lineno
        self.col_offset = col_offset


class Str(AST):
    __slots__ = ('s', 'lineno', 'col_offset')

    def __init__(self, s, lineno, col_offset):
        self.s = s
        self.lineno = lineno
        self.col_offset = col_offset


class Char(AST):
    __slots__ = ('c', 'lineno', 'col_offset')

    def __init__(self, c, lineno, col_offset):
        self.c = c
        self.lineno = lineno
        self.col_offset = col_offset


class Subscript(AST):
    __slots__ = ('name', 'slice', 'ctx', 'lineno', 'col_offset')

    def __init__(self, name, slice, ctx, lineno, col_offset):
        self.name = name
        self.slice = slice
        self.ctx = ctx
        self.lineno = lineno
        self.col_offset = col_offset


class Name(AST):
    __slots__ = ('id', 'ctx', 'lineno', 'col_offset')

    def __init__(self, id, ctx, lineno, col_offset):
        self.id = id
        self.ctx = ctx
        self.lineno = lineno
        self.col_offset = col_offset


class VarType(AST):
    __slots__ = ('type', 'is_array', 'size')

    def __init__(self, type, is_array, size):
        self.type = type
        self.is_array = is_array
        self.size = size



string2operator = {
    "+": operator.Add,
    "-": operator.Sub,
    "*": operator.Mult,
    "/": operator.Div,
    "==": operator.Eq,
    "!=": operator.NotEq,
    "<": operator.Lt,
    "<=": operator.LtE,
    ">": operator.Gt,
    ">=": operator.GtE,
}

string2unaryop = {
    "+": unaryop.UAdd,
    "-": unaryop.USub,
}

string2basic_type = {
    "int": basic_type.Int,
    "char": basic_type.Character,
    "void": basic_type.Void,
}

File: Analyzer/AST.cpp

#include "AST.h"

template <class T>
std::ostream &operator<<(std::ostream &os, const std::vector<T> &v) {
  os << "[";
  for (auto b = v.begin(), e = v.end(); b != e; ++b) {
    os << *b;
    if (b != e - 1) {
      os << ", ";
    }
  }
  os << "]";
  return os;
}

template <class T>
std::ostream &operator<<(std::ostream &os, const std::optional<T> &v) {
  if (v.has_value())
    return os << v.value();
  else
    return os << "None";
}

void Program::Format(std::ostream &os) const {
  os << "Program("
     << "decls=" << decls << ")";
}

void ConstDecl::Format(std::ostream &os) const {
  os << "ConstDecl("
     << "type=" << type << ", "
     << "name=" << name << ", "
     << "value=" << value << ")";
}

void VarDecl::Format(std::ostream &os) const {
  os << "VarDecl("
     << "type=" << type << ", "
     << "is_array=" << is_array << ", "
     << "size=" << size << ", "
     << "name=" << name << ")";
}

void FuncDef::Format(std::ostream &os) const {
  os << "FuncDef("
     << "return_type=" << return_type << ", "
     << "name=" << name << ", "
     << "args=" << args << ", "
     << "decls=" << decls << ", "
     << "stmts=" << stmts << ")";
}

void Arg::Format(std::ostream &os) const {
  os << "Arg("
     << "type=" << type << ", "
     << "name=" << name << ", "
     << "loc=" << loc << ")";
}

void Read::Format(std::ostream &os) const {
  os << "Read("
     << "names=" << names << ")";
}

void Write::Format(std::ostream &os) const {
  os << "Write("
     << "str=" << str << ", "
     << "value=" << value << ")";
}

void Assign::Format(std::ostream &os) const {
  os << "Assign("
     << "target=" << target << ", "
     << "value=" << value << ")";
}

void For::Format(std::ostream &os) const {
  os << "For("
     << "initial=" << initial << ", "
     << "condition=" << condition << ", "
     << "step=" << step << ", "
     << "body=" << body << ")";
}

void While::Format(std::ostream &os) const {
  os << "While("
     << "condition=" << condition << ", "
     << "body=" << body << ")";
}

void Return::Format(std::ostream &os) const {
  os << "Return("
     << "value=" << value << ")";
}

void If::Format(std::ostream &os) const {
  os << "If("
     << "test=" << test << ", "
     << "body=" << body << ", "
     << "orelse=" << orelse << ")";
}

void ExprStmt::Format(std::ostream &os) const {
  os << "ExprStmt("
     << "value=" << value << ")";
}

void BinOp::Format(std::ostream &os) const {
  os << "BinOp("
     << "left=" << left << ", "
     << "op=" << op << ", "
     << "right=" << right << ")";
}

void UnaryOp::Format(std::ostream &os) const {
  os << "UnaryOp("
     << "op=" << op << ", "
     << "operand=" << operand << ")";
}

void Call::Format(std::ostream &os) const {
  os << "Call("
     << "func=" << func << ", "
     << "args=" << args << ")";
}

void Num::Format(std::ostream &os) const {
  os << "Num("
     << "n=" << n << ")";
}

void Str::Format(std::ostream &os) const {
  os << "Str("
     << "s=" << s << ")";
}

void Char::Format(std::ostream &os) const {
  os << "Char("
     << "c=" << c << ")";
}

void Subscript::Format(std::ostream &os) const {
  os << "Subscript("
     << "name=" << name << ", "
     << "index=" << index << ", "
     << "ctx=" << ctx << ")";
}

void Name::Format(std::ostream &os) const {
  os << "Name("
     << "id=" << id << ", "
     << "ctx=" << ctx << ")";
}

std::ostream &operator<<(std::ostream &os, OperatorKind val) {
  switch (val) {

  case OperatorKind::Add:
    return os << "OperatorKind::Add";

  case OperatorKind::Sub:
    return os << "OperatorKind::Sub";

  case OperatorKind::Mult:
    return os << "OperatorKind::Mult";

  case OperatorKind::Div:
    return os << "OperatorKind::Div";

  case OperatorKind::Eq:
    return os << "OperatorKind::Eq";

  case OperatorKind::NotEq:
    return os << "OperatorKind::NotEq";

  case OperatorKind::Lt:
    return os << "OperatorKind::Lt";

  case OperatorKind::LtE:
    return os << "OperatorKind::LtE";

  case OperatorKind::Gt:
    return os << "OperatorKind::Gt";

  case OperatorKind::GtE:
    return os << "OperatorKind::GtE";
  }
}

std::ostream &operator<<(std::ostream &os, UnaryopKind val) {
  switch (val) {

  case UnaryopKind::UAdd:
    return os << "UnaryopKind::UAdd";

  case UnaryopKind::USub:
    return os << "UnaryopKind::USub";
  }
}

std::ostream &operator<<(std::ostream &os, ExprContextKind val) {
  switch (val) {

  case ExprContextKind::Load:
    return os << "ExprContextKind::Load";

  case ExprContextKind::Store:
    return os << "ExprContextKind::Store";
  }
}

std::ostream &operator<<(std::ostream &os, BasicTypeKind val) {
  switch (val) {

  case BasicTypeKind::Int:
    return os << "BasicTypeKind::Int";

  case BasicTypeKind::Character:
    return os << "BasicTypeKind::Character";

  case BasicTypeKind::Void:
    return os << "BasicTypeKind::Void";
  }
}

Program::~Program() {
  for (auto v : decls)
    delete v;
}

ConstDecl::~ConstDecl() { delete value; }

VarDecl::~VarDecl() {}

FuncDef::~FuncDef() {
  for (auto v : args)
    delete v;
  for (auto v : decls)
    delete v;
  for (auto v : stmts)
    delete v;
}

Arg::~Arg() {}

Read::~Read() {
  for (auto v : names)
    delete v;
}

Write::~Write() {
  if (str)
    delete str;
  if (value)
    delete value;
}

Assign::~Assign() {
  delete target;
  delete value;
}

For::~For() {
  delete initial;
  delete condition;
  delete step;
  for (auto v : body)
    delete v;
}

While::~While() {
  delete condition;
  for (auto v : body)
    delete v;
}

Return::~Return() {
  if (value)
    delete value;
}

If::~If() {
  delete test;
  for (auto v : body)
    delete v;
  for (auto v : orelse)
    delete v;
}

ExprStmt::~ExprStmt() { delete value; }

BinOp::~BinOp() {
  delete left;
  delete right;
}

UnaryOp::~UnaryOp() { delete operand; }

Call::~Call() {
  for (auto v : args)
    delete v;
}

Num::~Num() {}

Str::~Str() {}

Char::~Char() {}

Subscript::~Subscript() { delete index; }

Name::~Name() {}

OperatorKind String2OperatorKind(const String &s) {

  if (s == "+")
    return OperatorKind::Add;

  if (s == "-")
    return OperatorKind::Sub;

  if (s == "*")
    return OperatorKind::Mult;

  if (s == "/")
    return OperatorKind::Div;

  if (s == "==")
    return OperatorKind::Eq;

  if (s == "!=")
    return OperatorKind::NotEq;

  if (s == "<")
    return OperatorKind::Lt;

  if (s == "<=")
    return OperatorKind::LtE;

  if (s == ">")
    return OperatorKind::Gt;

  if (s == ">=")
    return OperatorKind::GtE;

  assert(false && "not a member of OperatorKind");
}

UnaryopKind String2UnaryopKind(const String &s) {

  if (s == "+")
    return UnaryopKind::UAdd;

  if (s == "-")
    return UnaryopKind::USub;

  assert(false && "not a member of UnaryopKind");
}

BasicTypeKind String2BasicTypeKind(const String &s) {

  if (s == "int")
    return BasicTypeKind::Int;

  if (s == "char")
    return BasicTypeKind::Character;

  if (s == "void")
    return BasicTypeKind::Void;

  assert(false && "not a member of BasicTypeKind");
}

String GetDeclName(Decl *decl) {
  if (auto x = subclass_cast<ConstDecl>(decl)) {
    return x->name;
  } else if (auto x = subclass_cast<VarDecl>(decl)) {
    return x->name;
  } else {
    auto y = subclass_cast<FuncDef>(decl);
    assert(y);
    return y->name;
  }
}
